70_FLIPFluids/app_resources/compute/radix_sort/radixSort.comp.hlsl

```diff
@@ -0,0 +1,111 @@
+#include "sort_common.hlsl"
+
+using namespace nbl::hlsl;
+
+#define GET_KEY(s) s.x
+
+[[vk::binding(0, 1)]]
+cbuffer SortParams
+{
+    SSortParams params;
+};
+
+[[vk::binding(1, 1)]] RWStructuredBuffer<DATA_TYPE> inputBuffer;
+[[vk::binding(2, 1)]] RWStructuredBuffer<DATA_TYPE> outputBuffer;
+
+[[vk::binding(3, 1)]] RWStructuredBuffer<uint> histograms;
+
+groupshared uint sums[NumSortBins / SubgroupSize];
+groupshared uint globalOffsets[NumSortBins];
+
+struct BinFlags
+{
+    uint flags[WorkgroupSize / 32];
+};
+groupshared BinFlags binFlags[NumSortBins];
+
+[numthreads(WorkgroupSize, 1, 1)]
+void main(uint threadID : SV_GroupThreadID, uint groupID : SV_GroupID)
+{
+    uint g_id = groupID.x;
+    uint l_id = threadID.x;
+    uint s_id = glsl::gl_SubgroupID();
+    uint ls_id = glsl::gl_SubgroupInvocationID();
+
+    uint localHistogram = 0;
+    uint prefixSum = 0;
+    uint histogramCount = 0;
+
+    if (l_id < NumSortBins)
+    {
+        uint count = 0;
+        for (uint i = 0; i < params.numWorkgroups; i++)
+        {
+            const uint t = histograms[NumSortBins * i + l_id];
+            localHistogram = (i == g_id) ? count : localHistogram;
+            count += t;
+        }
+
+        histogramCount = count;
+        const uint sum = glsl::subgroupAdd(histogramCount);
+        prefixSum = glsl::subgroupExclusiveAdd(histogramCount);
+        if (glsl::subgroupElect())
+            sums[s_id] = sum;
+    }
+    glsl::barrier();
+
+    if (l_id < NumSortBins)
+    {
+        const uint totalPrefixSums = glsl::subgroupBroadcast(glsl::subgroupExclusiveAdd(sums[ls_id]), s_id);
+        const uint globalHistogram = totalPrefixSums + prefixSum;
+        globalOffsets[l_id] = globalHistogram + localHistogram;
+    }
+
+    const uint flagsBin = l_id / 32;
+    const uint flagsBit = 1 << (l_id % 32);
+
+    for (uint i = 0; i < params.numThreadsPerGroup; i++)
+    {
+        uint elementID = g_id * params.numThreadsPerGroup * WorkgroupSize + i * WorkgroupSize + l_id;
+
+        if (l_id < NumSortBins)
+        {
+            for (int j = 0; j < WorkgroupSize / 32; j++)
```

Comment:
don't use unnamed literal constants, I have no clue if the 32 is the number of bits in a uint32_t or you hardcoded the SubgroupSize according to Nvidia

Reply:
oof I find I'm repeating myself